gridfs.go

```diff
@@ -0,0 +1,601 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "opening stream")
+	}
+
+	return writer, nil
+}
+
+func (b *gridfsBucket) Reader(ctx context.Context, name string) (io.ReadCloser, error) {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"operation":     "reader",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+	})
+
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "resolving bucket")
+	}
+
+	reader, err := grid.OpenDownloadStreamByName(b.normalizeKey(name))
+	if err != nil {
+		if err == gridfs.ErrFileNotFound {
+			err = MakeKeyNotFoundError(err)
+		}
+		return nil, errors.Wrap(err, "opening stream")
+	}
+
+	return reader, nil
+}
+
+func (b *gridfsBucket) Put(ctx context.Context, name string, input io.Reader) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"dry_run":       b.opts.DryRun,
+		"operation":     "put",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+	})
+
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return errors.Wrap(err, "resolving bucket")
+	}
+
+	if b.opts.DryRun {
+		return nil
+	}
+
+	if _, err = grid.UploadFromStream(b.normalizeKey(name), input); err != nil {
+		return errors.Wrap(err, "uploading file")
+	}
+
+	return nil
+}
+
+func (b *gridfsBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"operation":     "get",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+	})
+
+	return b.Reader(ctx, name)
+}
+
+func (b *gridfsBucket) Upload(ctx context.Context, name, path string) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"dry_run":       b.opts.DryRun,
+		"operation":     "upload",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+		"path":          path,
+	})
+
+	f, err := os.Open(path)
+	if err != nil {
+		return errors.Wrapf(err, "opening file '%s'", name)
+	}
+	defer f.Close()
+
+	return errors.WithStack(b.Put(ctx, name, f))
+}
+
+func (b *gridfsBucket) Download(ctx context.Context, name, path string) error {
+	grip.DebugWhen(b.opts.Verbose, message.Fields{
+		"type":          "gridfs",
+		"operation":     "download",
+		"bucket":        b.opts.Name,
+		"bucket_prefix": b.opts.Prefix,
+		"key":           name,
+		"path":          path,
+	})
+
+	reader, err := b.Reader(ctx, name)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err = os.MkdirAll(filepath.Dir(path), 0700); err != nil {
+		return errors.Wrapf(err, "creating enclosing directory for file '%s'", path)
+	}
+
+	f, err := os.Create(path)
+	if err != nil {
+		return errors.Wrapf(err, "creating file '%s'", path)
+	}
+	_, err = io.Copy(f, reader)
+	if err != nil {
+		_ = f.Close()
```

Comment:
Seems kind of weird to ignore the file close error here, but return it if the copy succeeds. Should this just `defer f.Close()` above, or should we make this consistent and return the two errors accumulated in a catcher?

Reply:
I think we can just do a `defer f.Close()` above. Do we really care about those errors? In most other places we just do a defer and ignore them.

Reply:
I don't think closing a file flushes so it wouldn't matter from a user perspective.

Reply:
Just defer closing sounds good.